### PR TITLE
HHH-16733 HHH-16582 QuerySplitter fixes for function parameters and joins

### PR DESCRIPTION
**Pull request by mbladel**
https://hibernate.atlassian.net/browse/HHH-16733
https://hibernate.atlassian.net/browse/HHH-16582

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/query/hql/internal/QuerySplitter.java [hunk 1]
```diff
@@ -736,13 +764,14 @@ public SqmCorrelatedRoot<?> visitCorrelatedRoot(SqmCorrelatedRoot<?> correlatedR
 
 		@Override
 		public SqmBagJoin<?, ?> visitBagJoin(SqmBagJoin<?, ?> join) {
+			final SqmFrom lhsCopy = findSqmFromCopy( join.getLhs() );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/query/hql/internal/QuerySplitter.java:767]: Maybe I'm not seeing it, but why did you move the `findSqmFromCopy` call up here?

**Review by beikov — APPROVED** [review]

File: hibernate-core/src/main/java/org/hibernate/query/hql/internal/QuerySplitter.java (hunk 1 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/query/hql/internal/QuerySplitter.java:767]: Because we could copy a join with a lhs that was still not copied (e.g. another join), thus triggering the copy of that as well. That would in turn trigger a new copy of the original join (since that was still being constructed and not yet registered) thus duplicating the object and triggering an "Alias [x] used for multiple from-clause elements" error.

Putting `findSqmFromCopy` of the lhs before checking the `sqmFromCopyMap` allows us to always only copy the join once and reuse the instance correctly.

Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/query/hql/internal/QuerySplitter.java:767]: Ok, I guess the name `findSqmFromCopy` confused me 😅 
Would you mind trying to use the following implementation for `split` instead?

```
for ( EntityDomainType<?> mappedDescriptor : unmappedPolymorphicDescriptor.getImplementors() ) {
	i++;
	final SqmCopyContext copyContext = SqmCopyContext.simpleContext();
	copyContext.registerCopy( 
		unmappedPolymorphicFromElement,
		new SqmRoot<>(
					mappedDescriptor,
					unmappedPolymorphicFromElement.getExplicitAlias(),
					unmappedPolymorphicFromElement.isAllowJoins(),
					unmappedPolymorphicFromElement.nodeBuilder()
		)
 	);
	expanded[i] = statement.copy( copyContext );
}
```

I think this will allow us to get rid of the entire `UnmappedPolymorphismReplacer`.

Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/query/hql/internal/QuerySplitter.java]: I've updated the PR to use the existing SQM `copy` logic as suggested, avoiding the copy of parameters.

I've also removed the need for an `instanceof SqmPath` check in the copy context by always copying SqmFrom's `lhs` before checking if the current object was already copied. Let me know what you think.

**Review by beikov — APPROVED** [review]

File: hibernate-core/src/main/java/org/hibernate/query/sqm/tree/domain/AbstractSqmPath.java [hunk 2]
```diff
@@ -212,6 +214,28 @@ protected <S extends T> SqmTreatedPath<T, S> getTreatedPath(EntityDomainType<S>
 		return path;
 	}
 
+	/**
+	 * Utility that checks if this path's parent navigable path is compatible with the specified SQM parent,
+	 * and if not creates a copy of the navigable path with the correct parent.
+	 */
+	protected NavigablePath getNavigablePathCopy(SqmPath<?> parent) {
+		final NavigablePath parentNavigablePath = NullnessUtil.castNonNull( navigablePath.getRealParent() );
+		if ( parent.getNavigablePath().pathsMatch( parentNavigablePath ) ) {
```
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/query/sqm/tree/domain/AbstractSqmPath.java:223]: Yeah this is what I initially tried, but I had a couple failures throughout the test suite: when we have correlated paths, the `==` check fails but the navigable path is effectively correct (i.e. the actual path is the same) while we have different `NavigablePath` instances.

This, in turn, was causing the creation of a new navigable path object and thus a failure to locate the correct `TableGroup` for the parent path.

**Review by beikov — APPROVED** [review]
